File: chad/src/main/java/org/woahoverflow/chad/core/listener/GuildJoinLeave.java
package org.woahoverflow.chad.core.listener;

import org.woahoverflow.chad.framework.handle.GuildHandler;
import org.woahoverflow.chad.framework.handle.PlayerHandler;
import org.woahoverflow.chad.framework.handle.database.DatabaseManager;
import org.woahoverflow.chad.framework.obj.Guild;
import org.woahoverflow.chad.framework.obj.Player;
import org.woahoverflow.chad.framework.obj.Player.DataType;
import org.woahoverflow.chad.framework.ui.UI;
import sx.blah.discord.api.events.EventSubscriber;
import sx.blah.discord.handle.impl.events.guild.GuildCreateEvent;
import sx.blah.discord.handle.impl.events.guild.GuildLeaveEvent;
import sx.blah.discord.handle.obj.IChannel;
import sx.blah.discord.handle.obj.IUser;
import sx.blah.discord.handle.obj.Permissions;
import sx.blah.discord.util.RequestBuffer;

import java.util.ArrayList;
import java.util.List;

/**
 * The Discord guild join and leave events
 *
 * @author sho, codebasepw
 */
public final class GuildJoinLeave {
    /**
     * Discord's Joining Guild Event
     *
     * @param event Guild Create Event
     */
    @EventSubscriber
    @SuppressWarnings({"unused", "unchecked"})
    public void joinGuild(GuildCreateEvent event) {
        // Makes sure all users are into the database
        new Thread(() -> {
            List<IUser> users = RequestBuffer.request(() -> event.getGuild().getUsers()).get();
            for (IUser user : users) {
                Player player = PlayerHandler.handle.getPlayer(user.getLongID());
                ArrayList<Long> guildData = (ArrayList<Long>) player.getObject(DataType.GUILD_DATA);
                if (!guildData.contains(event.getGuild().getLongID())) {
                    guildData.add(event.getGuild().getLongID());
                    player.setObject(DataType.GUILD_DATA, guildData);
                }
            }
        });

        if (!GuildHandler.handle.guildDataExists(event.getGuild().getLongID())) {
            // By retrieving the guild's instance, it creates an instance for the guild within the database
            Guild guild = GuildHandler.handle.getGuild(event.getGuild().getLongID());

            // Display the new guild in the UI
            UI.displayGuild(event.getGuild());

            // Send a log with the new guild
            UI.handle.addLog('[' +event.getGuild().getStringID()+"] Joined Guild", UI.LogLevel.INFO);

            // The guild's default channel
            IChannel defaultChannel = RequestBuffer.request(() -> event.getGuild().getDefaultChannel()).get();

            // The join message
            final String joinMessage = "Hello, i'm Chad!\n"
                + "My prefix is `c!`\n\n"
                + "View my commands with `c!help`";

            // If the bot has permission to, send the join message into the default channel
            if (RequestBuffer.request(() -> defaultChannel.getModifiedPermissions(event.getClient().getOurUser()).contains(Permissions.SEND_MESSAGES)).get())
                RequestBuffer.request(() -> event.getGuild().getDefaultChannel().sendMessage(joinMessage));
            else {
                // Parse through all of the guilds, and if the bot has permission send the join message.
                List<IChannel> guilds = RequestBuffer.request(() -> event.getGuild().getChannels()).get();
                int channelSize = guilds.size();

                for (int i = 0; channelSize > i; i++) {
                    IChannel channel = guilds.get(0);

                    if (RequestBuffer.request(() -> channel.getModifiedPermissions(event.getClient().getOurUser()).contains(Permissions.SEND_MESSAGES)).get()) {
                        RequestBuffer.request(() -> channel.sendMessage(joinMessage));
                    }
                }
            }
        }
    }

    /**
     * Discord's Leave Guild Event
     *
     * @param event Guild Leave Event
     */
    @EventSubscriber
    @SuppressWarnings("unused")
    public void leaveGuild(GuildLeaveEvent event)
    {
        // Delete the guild's document
        DatabaseManager.GUILD_DATA.removeDocument(event.getGuild().getStringID());

        // Removed the guild's cached document
        GuildHandler.handle.removeGuild(event.getGuild().getLongID());

        // Send a log
        UI.handle.addLog('<' +event.getGuild().getStringID()+"> Left Guild", UI.LogLevel.INFO);
    }
}

File: chad/src/main/java/org/woahoverflow/chad/core/listener/GuildJoinLeave.kt
package org.woahoverflow.chad.core.listener

import org.woahoverflow.chad.core.ChadInstance
import org.woahoverflow.chad.framework.handle.GuildHandler
import org.woahoverflow.chad.framework.handle.PlayerHandler
import org.woahoverflow.chad.framework.obj.Player.DataType
import sx.blah.discord.api.events.EventSubscriber
import sx.blah.discord.handle.impl.events.guild.GuildCreateEvent
import sx.blah.discord.handle.impl.events.guild.GuildLeaveEvent
import sx.blah.discord.handle.obj.IChannel
import sx.blah.discord.handle.obj.IMessage
import sx.blah.discord.handle.obj.IUser
import sx.blah.discord.handle.obj.Permissions
import sx.blah.discord.util.RequestBuffer
import java.util.*

/**
 * The Discord guild join and leave events
 *
 * @author sho, codebasepw
 */
class GuildJoinLeave {
    /**
     * Discord's Joining Guild Event
     *
     * @param event Guild Create Event
     */
    @EventSubscriber
    fun joinGuild(event: GuildCreateEvent) {
        // Makes sure all users are into the database
        Thread {
            val users = RequestBuffer.request<List<IUser>> { event.guild.users }.get()
            for (user in users) {
                val player = PlayerHandler.getPlayer(user.longID)
                val guildData = player.getObject(DataType.GUILD_DATA) as ArrayList<Long>

                if (!guildData.contains(event.guild.longID)) {
                    guildData.add(event.guild.longID)
                    player.setObject(DataType.GUILD_DATA, guildData)
                }
            }
        }.run()

        if (!GuildHandler.guildDataExists(event.guild.longID)) {
            // By retrieving the guild's instance, it creates an instance for the guild within the database
            GuildHandler.getGuild(event.guild.longID)

            // Send a log with the new guild
            ChadInstance.getLogger().debug('['.toString() + event.guild.stringID + "] Joined Guild")

            // The guild's default channel
            val defaultChannel = RequestBuffer.request<IChannel> { event.guild.defaultChannel }.get()

            // The join message
            val joinMessage = ("Hello, i'm Chad!\n"
                    + "My prefix is `c!`\n\n"
                    + "View my commands with `c!help`")

            // If the bot has permission to, send the join message into the default channel
            if (RequestBuffer.request<Boolean> { defaultChannel.getModifiedPermissions(event.client.ourUser).contains(Permissions.SEND_MESSAGES) }.get())
                RequestBuffer.request<IMessage> { event.guild.defaultChannel.sendMessage(joinMessage) }
            else {
                // Parse through all of the guilds, and if the bot has permission send the join message.
                val guilds = RequestBuffer.request<List<IChannel>> { event.guild.channels }.get()
                val channelSize = guilds.size

                var i = 0
                while (channelSize > i) {
                    val channel = guilds[0]

                    if (RequestBuffer.request<Boolean> { channel.getModifiedPermissions(event.client.ourUser).contains(Permissions.SEND_MESSAGES) }.get()) {
                        RequestBuffer.request<IMessage> { channel.sendMessage(joinMessage) }
                    }
                    i++
                }
            }
        }
    }

    /**
     * Discord's Leave Guild Event
     *
     * @param event Guild Leave Event
     */
    @EventSubscriber
    fun leaveGuild(event: GuildLeaveEvent) {
        // Removed the guild's cached document
        GuildHandler.removeGuild(event.guild.longID)

        // Send a log
        ChadInstance.getLogger().debug('<'.toString() + event.guild.stringID + "> Left Guild")
    }
}
